src/lib/device/journal_vdev.hpp

```diff
@@ -19,250 +19,367 @@
 #include <atomic>
 #include <functional>
 #include <memory>
+#include <vector>
+#include <condition_variable>
 
 #include "device.h"
+#include "physical_dev.hpp"
 #include "virtual_dev.hpp"
 
 namespace homestore {
 typedef std::function< void(const off_t ret_off) > alloc_next_blk_cb_t;
+using journal_id_t = uint64_t;
+// Each logstore family is associated to a single logdevice.
+using logdev_id_t = uint64_t;
+
+// Chunks used for journal vdev has journal related info stored in chunk private data.
+// Each log device has a list of journal chunk data with next_chunk.
+// Journal vdev will arrange the chunks in order during recovery.
+struct JournalChunkPrivate {
+    logdev_id_t logdev_id{0};
+    bool is_head{false};           // Is it the head element.
+    uint64_t created_at{0};        // Creation timestamp
+    uint64_t end_of_chunk_size{0}; // The offset indicates end of chunk.
+    chunk_num_t next_chunk{0};     // Next chunk in the list.
+};
+
+static_assert(sizeof(JournalChunkPrivate) <= chunk_info::user_private_size, "Journal private area bigger");
 
 class JournalVirtualDev : public VirtualDev {
-    struct Chunk_EOF_t {
-        uint64_t e;
+public:
+    // Client use journal vdev open to create a descriptor to append log entries.
+    // Each descriptor is independent list of chunks in order and like sliding window
+    // maintains size, offsets like the left side (m_data_start_offset) and
+    // right side (m_end_offset). Truncate increases the left m_data_start_offset
+    // and pop chunks from the front of the list. alloc_next_append_blk adds more chunk to
+    // the back of list if no space and adjusts the right m_end_offset. All offsets
+    // only increase and never wraps around. Each chunk in the descriptor has private
+    // data about the logdev_id its part of, next chunk to maintain the list.
+    struct Descriptor {
+    private:
+        JournalVirtualDev& m_vdev;
+        logdev_id_t m_logdev_id; // Unique id identifying the journal descriptor.
+        // off_t is long. make it uint64_t ?
```

Comment:
good point, long is only 4 bytes, and it is signed, it will overflow if exceeding 2TB offset, right?

Reply:
Currently off_t is mapped to int64_t which itself is huge.